pkg/daemon/ceph/cleanup/subvolumegroups.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2024 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+)
+
+func SubVolumeGroupCleanup(context *clusterd.Context, clusterInfo *client.ClusterInfo, fsName, svg, poolName, csiNamespace string) error {
+	logger.Infof("starting clean up cephFS subVolumeGroup resource %q", svg)
+
+	subVolumeList, err := cephclient.ListSubvolumesInGroup(context, clusterInfo, fsName, svg)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list cephFS subVolumes in subVolumeGroup %q", svg)
+	}
+
+	for _, subVolume := range subVolumeList {
+		logger.Infof("starting clean up of subvolume %q", subVolume.Name)
+		err := CleanUpOMAPDetails(context, clusterInfo, subVolume.Name, poolName, csiNamespace)
+		if err != nil {
+			logger.Errorf("failed to clean up OMAP details for the subvolume %q. %v", subVolume.Name, err)
+		}
+
+		subVolumeSnapshots, err := cephclient.ListSubVolumeSnapshots(context, clusterInfo, fsName, subVolume.Name, svg)
+		if err != nil {
+			logger.Errorf("failed to list snapshots for subvolume %q in group %q. %v", subVolume.Name, svg, err)
+		} else {
+			err := CancelPendingClones(context, clusterInfo, subVolumeSnapshots, fsName, subVolume.Name, svg)
+			if err != nil {
+				logger.Errorf("failed to cancel pending clones for subvolume %q in group %q. %v", subVolume.Name, svg, err)
+			}
+			err = DeleteSubVolumeSnapshots(context, clusterInfo, subVolumeSnapshots, fsName, subVolume.Name, svg)
+			if err != nil {
+				logger.Errorf("failed to delete snapshots for subvolume %q in group %q. %v", subVolume.Name, svg, err)
+			}
+		}
+
+		err = cephclient.DeleteSubVolume(context, clusterInfo, fsName, subVolume.Name, svg)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete subvolume Group %q.", subVolume.Name)
```

Comment:
Instead of returning immediately on failure, store the error in a var and attempt deletion of all subvolumes. Then at the end return the error if any. This way, we can always attempt to cleanup as much as possible.